File: polyglot/csharp/disk_image_processing.cs
using System;
using System.IO;
using System.Collections.Generic;

namespace polyglot.csharp
{
    class DiskImageProcessor
    {
        // Represents a file signature with its name and byte pattern
        private class FileSignature
        {
            public string Name { get; }
            public byte[] Pattern { get; }
            public int Length { get; }

            public FileSignature(string name, byte[] pattern)
            {
                Name = name;
                Pattern = pattern;
                Length = pattern.Length;
            }
        }

        // List of known file signatures
        private static readonly List<FileSignature> KnownSignatures = new List<FileSignature>
        {
            new FileSignature("JPEG", new byte[] { 0xFF, 0xD8, 0xFF, 0xE0, 0x00, 0x10, 0x4A, 0x45, 0x46, 0x49, 0x63, 0x00 }),
            new FileSignature("PNG", new byte[] { 0x89, 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A }),
            new FileSignature("ZIP", new byte[] { 0x50, 0x4B, 0x03, 0x04 }),
            new FileSignature("PDF", new byte[] { 0x25, 0x50, 0x44, 0x46 }),
            new FileSignature("TXT", new byte[] { 0x48, 0x65, 0x6C, 0x6C, 0x6F, 0x20, 0x57, 0x6F, 0x72, 0x6C, 0x64 }),
            new FileSignature("MP3", new byte[] { 0x49, 0x44, 0x33, 0x03 }),
            new FileSignature("AVI", new byte[] { 0x41, 0x56, 0x49, 0x20, 0x33, 0x30, 0x31, 0x30 })
        };

        // Carve files from a disk image by known signatures
        public static void CarveFiles(string imagePath, string outputDir)
        {
            if (!Directory.Exists(outputDir))
                Directory.CreateDirectory(outputDir);

            using (var reader = new BinaryReader(File.OpenRead(imagePath)))
            {
                int offset = 0;
                while (reader.BaseStream.Position < reader.BaseStream.Length)
                {
                    foreach (var signature in KnownSignatures)
                    {
                        if (reader.BaseStream.Position + signature.Length > reader.BaseStream.Length)
                            break;

                        byte[] buffer = new byte[signature.Length];
                        reader.Read(buffer, 0, buffer.Length);

                        if (CompareByteArrays(buffer, signature.Pattern))
                        {
                            string fileName = $"{signature.Name}_{offset}.carved";
                            string filePath = Path.Combine(outputDir, fileName);

                            // Carve the file by reading ahead until the end of the file
                            int fileSize = 0;
                            byte[] fileBuffer = new byte[4096];
                            using (var writer = new BinaryWriter(File.Create(filePath)))
                            {
                                while (true)
                                {
                                    int bytesRead = reader.Read(fileBuffer, 0, fileBuffer.Length);
                                    if (bytesRead == 0) break;

                                    writer.Write(fileBuffer, 0, bytesRead);
                                    fileSize += bytesRead;

                                    // Check for end of file signature or EOF
                                    if (fileSize > 1024 * 1024 * 10) // Limit to 10MB to prevent infinite loops
                                        break;
                                }
                            }

                            Console.WriteLine($"Found {signature.Name} at offset {offset} (size: {fileSize} bytes)");
                        }
                    }

                    offset += 1;
                }
            }
        }

        // Compare two byte arrays for equality
        private static bool CompareByteArrays(byte[] a, byte[] b)
        {
            if (a.Length != b.Length) return false;

            for (int i = 0; i < a.Length; i++)
            {
                if (a[i] != b[i]) return false;
            }

            return true;
        }

        // Main entry point with demo
        public static void Main(string[] args)
        {
            string imagePath = "sample_disk_image.dd";
            string outputDir = "carved_files";

            Console.WriteLine("Starting disk image processing...");
            CarveFiles(imagePath, outputDir);
            Console.WriteLine("Processing complete.");
        }
    }
}

File: polyglot/c/disk_image_processing.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/stat.h>
#include <sys/types.h>

#define MAX_SIGNATURES 10
#define MAX_FILE_NAME 256
#define BUFFER_SIZE 4096

typedef struct {
    char *signature;
    char *file_type;
} SignatureEntry;

// Sample signatures for common file types
SignatureEntry signatures[] = {
    {"\x50\x4b\x03\x04", "ZIP"},
    {"\xD0\xCF\x11\xE0\xA1\xB1\x1A\xE1", "MS Office (DOCX)"},
    {"\x25\x50\x44\x46", "PDF"},
    {"\x49\x49\x2A\x00", "TIFF"},
    {"\x42\x4D", "Windows Bitmap"},
    {"\x52\x61\x72\x21\x19\x00\x01\x00", "RAR"},
    {"\x47\x49\x46\x38", "GIF"},
    {"\x49\x49\x43\x4D", "PNG"},
    {"\xFF\xD8\xFF", "JPEG"},
    {"\x50\x4B\x01\x02", "ZIP (streamed)"}
};

int find_signature(const unsigned char *buffer, size_t size) {
    for (int i = 0; i < MAX_SIGNATURES; i++) {
        if (size >= strlen(signatures[i].signature)) {
            if (memcmp(buffer, signatures[i].signature, strlen(signatures[i].signature)) == 0) {
                return i;
            }
        }
    }
    return -1;
}

void carve_files_from_image(const char *image_path, const char *output_dir) {
    int fd = open(image_path, O_RDONLY);
    if (fd == -1) {
        perror("Failed to open image");
        return;
    }

    char buffer[BUFFER_SIZE];
    ssize_t bytes_read;
    char file_name[MAX_FILE_NAME];
    int file_count = 0;

    while ((bytes_read = read(fd, buffer, sizeof(buffer))) > 0) {
        int sig_idx = find_signature(buffer, bytes_read);
        if (sig_idx != -1) {
            snprintf(file_name, MAX_FILE_NAME, "%s_%d.%s", output_dir, file_count++, signatures[sig_idx].file_type);
            FILE *out_file = fopen(file_name, "wb");
            if (!out_file) {
                perror("Failed to create output file");
                continue;
            }

            // Write the matched signature and subsequent data
            fwrite(buffer, 1, bytes_read, out_file);
            fclose(out_file);
        }
    }

    close(fd);
    printf("Carved %d files.\n", file_count);
}

int main(int argc, char *argv[]) {
    if (argc != 3) {
        fprintf(stderr, "Usage: %s <image_path> <output_dir>\n", argv[0]);
        return 1;
    }

    const char *image_path = argv[1];
    const char *output_dir = argv[2];

    // Create output directory if it doesn't exist
    if (mkdir(output_dir, 0777) == -1 && errno != EEXIST) {
        perror("Failed to create output directory");
        return 1;
    }

    carve_files_from_image(image_path, output_dir);

    return 0;
}